### PR TITLE
[Merged by Bors] - Implement slashing protection interchange format

### LINKED ISSUE #1584: Register validators with slashing protection as they are created/imported
## Description

The validator client should not automatically register validators with the slashing protection database, as this poses a slashing risk if an existing slashing protection database is misplaced.

A validator should only need to be registered with slashing protection once -- when it first comes to exist, either through `lighthouse account validator create` or `lighthouse account validator import`.

## Version

Lighthouse v0.2.8-967700c
BLS Library: blst

### PR DESCRIPTION
**Pull request by michaelsproul**
## Issue Addressed

Implements support for importing and exporting the slashing protection DB interchange format described here:

https://hackmd.io/@sproul/Bk0Y0qdGD

Also closes #1584 

## Proposed Changes

* [x] Support for serializing and deserializing the format
* [x] Support for importing and exporting Lighthouse's database
* [x] CLI commands to invoke import and export
* [x] Export to minimal format (required when a minimal format has been previously imported)
* [x] Tests for export to minimal (utilising mixed importing and attestation signing?)
* [x] Tests for import/export of complete format, and import of minimal format
* [x] ~~Prevent attestations with sources less than our max source (Danny's suggestion). Required for the fake attestation that we put in for the minimal format to block attestations from source 0.~~
* [x] Add the concept of a "low watermark" for compatibility with the minimal format

Bonus!

* [x] A fix to a potentially nasty bug involving validators getting re-registered each time the validator client ran! Thankfully, the ordering of keys meant that the validator IDs used for attestations and blocks remained stable -- otherwise we could have had some slashings on our hands! :scream:
* [x] Tests to confirm that this bug is indeed vanquished

### CODE REVIEW

**Comment by djrtwo** [comment]
Are you implementing both formats from the doc? Or should teku instead just use the first format and just provide one stubbed message for attestations and blocks

**Comment by michaelsproul** [comment]
@djrtwo Yeah I've already implemented both! I just need to write some tests and a CLI porcelain, but have been distracted by fixing some DB issues

**Comment by michaelsproul** [comment]
I'm planning to try a different approach to importing the minimal format. Rather than filling the DB with fake blocks for every slot from 0 to `last_signed_block_slot` (which is really slow), I want to add the last signed block slot as a first-class entity in slashing protection, which we can set with one query. This would also allow us to prune the DB, although I might not implement that yet.

A similar low watermark for attestations will obsolete the need for the `source >= max_source` check on attestations, and the "clever" attestation import strategy with an attestation spanning `0->last_signed_attestation_target_epoch`. I realised that this strategy was going to get even messier, because in addition to the attestation from `0->N` which blocks `i->j` for `i`, `j` in `[1..N]`, we need one with a source greater than 0 to block `0->j` for `j` in `1..N` that isn't mutually slashable with the `0->N` one. I think there's a hacky way to do it with 2 attestations, but the low watermark really seems simpler and less error prone.

Glossary:

* `i->j` is an attestation with source epoch `i` and target `j`

**Comment by michaelsproul** [comment]
Blocked on #1588

**Comment by michaelsproul** [comment]
This is ready for review!

There's one FIXME that depends on https://github.com/sigp/lighthouse/pull/1532, but that should be straight-forward to resolve once one of these PRs is merged.

**Comment by michaelsproul** [comment]
Marking this as WIP again, because it needs updating for v4 of the spec

**Comment by michaelsproul** [comment]
This is ready to roll again, just waiting on resolution of merge conflicts in #1532

**Comment by michaelsproul** [comment]
This PR closes an A0 issue (#1584), but the fix is relatively well-contained in https://github.com/sigp/lighthouse/pull/1544/commits/1dbd5379d4fa433fdcabf5606fae7ba4cea2af99 if we want to cherry-pick it and merge it before reviewing+merging this PR.

**Review by paulhauner — CHANGES_REQUESTED** [review]
Impressive, especially with all the test vectors and interop contribution!

I just have a few comments. I played around with it and it worked great, I had a couple of suggestions for logs that might reduce our support load.

File: validator_client/slashing_protection/src/slashing_database.rs
Comment (michaelsproul) [reply validator_client/slashing_protection/src/slashing_database.rs:658]: Yes!!

File: validator_client/slashing_protection/src/interchange.rs
Comment (michaelsproul) [reply validator_client/slashing_protection/src/interchange.rs:23]: Yeah, I think I'll delete it in v5 of the spec, but as it's a breaking change, I figured it could stay a bit longer

File: validator_client/slashing_protection/build.rs
Comment (michaelsproul) [reply validator_client/slashing_protection/build.rs:2]: Thanks!

File: account_manager/src/validator/slashing_protection.rs
Comment (michaelsproul) [reply account_manager/src/validator/slashing_protection.rs:102]: Done!

File: account_manager/src/validator/slashing_protection.rs
Comment (michaelsproul) [reply account_manager/src/validator/slashing_protection.rs:98]: Done!

File: lighthouse/tests/account_manager.rs
Comment (michaelsproul) [reply lighthouse/tests/account_manager.rs]: Done

**Comment by michaelsproul** [comment]
All comments addressed :tada:

**Review by paulhauner — APPROVED** [review]
Approved! :tada:

**Comment by michaelsproul** [comment]
Thanks!

bors r+

**Pull request closed without merging**
